### PR TITLE
Rule RS0036: Flag DiagnosticDescriptor instances which don't specify …

### LINKED ISSUE #1501: Write an analyzer rule to warn if custom tags argument is not provided to the DiagnsoticDescriptor ctor
#### Analyzer

DiagnosticDescriptorCreationAnalyzer

#### Details

Rule should be disabled by default and turned on for our own repos (Roslyn, Roslyn-analyzers, etc.)

### PR DESCRIPTION
**Pull request by Evangelink**
…the customTags argument.

Fix #1501

### CODE REVIEW

File: src/Roslyn.Diagnostics.Analyzers/Core/SpecifyDiagnosticDescriptorCustomTags.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Roslyn.Diagnostics.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class SpecifyDiagnosticDescriptorCustomTags : DiagnosticAnalyzer
```
Comment (Evangelink) [inline src/Roslyn.Diagnostics.Analyzers/Core/SpecifyDiagnosticDescriptorCustomTags.cs]: Let me know if you prefer another name.

File: src/Roslyn.Diagnostics.Analyzers/Core/RoslynDiagnosticsAnalyzersResources.resx [hunk 2]
```diff
@@ -252,4 +252,10 @@
   <data name="TestExportsShouldNotBeDiscoverableCodeFix" xml:space="preserve">
     <value>Apply PartNotDiscoverableAttribute</value>
   </data>
+  <data name="SpecifyDiagnosticDescriptorCustomTagsMessage" xml:space="preserve">
+    <value>The 'customTags' argument should always be specified.</value>
+  </data>
+  <data name="SpecifyDiagnosticDescriptorCustomTagsTitle" xml:space="preserve">
+    <value>Add 'customTags' argument</value>
```
Comment (Evangelink) [inline src/Roslyn.Diagnostics.Analyzers/Core/RoslynDiagnosticsAnalyzersResources.resx]: Shall it be "Add the 'customTags' argument"?

File: src/Roslyn.Diagnostics.Analyzers/Core/SpecifyDiagnosticDescriptorCustomTags.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                        if (objectCreation.Type.Equals(diagnosticDescriptorType) && IsMissingCustomTagsArgument(objectCreation.Arguments))
+                        {
+                            opContext.ReportDiagnostic(objectCreation.CreateDiagnostic(Rule));
+                        }
+                    }, OperationKind.ObjectCreation);
+                }
+            });
+        }
+
+        private static bool IsMissingCustomTagsArgument(ImmutableArray<IArgumentOperation> arguments) =>
```
Comment (Evangelink) [inline src/Roslyn.Diagnostics.Analyzers/Core/SpecifyDiagnosticDescriptorCustomTags.cs]: I am not sure whether this is the best way to handle the value passed as argument. Please advise for a better option if any.

File: src/Roslyn.Diagnostics.Analyzers/Core/SpecifyDiagnosticDescriptorCustomTags.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                if (compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(typeof(DiagnosticDescriptor).FullName, out var diagnosticDescriptorType))
+                {
+                    compilationContext.RegisterOperationAction(opContext =>
+                    {
+                        var objectCreation = (IObjectCreationOperation)opContext.Operation;
+                        if (objectCreation.Type.Equals(diagnosticDescriptorType) && IsMissingCustomTagsArgument(objectCreation.Arguments))
+                        {
+                            opContext.ReportDiagnostic(objectCreation.CreateDiagnostic(Rule));
+                        }
+                    }, OperationKind.ObjectCreation);
```
Comment (Evangelink) [inline src/Roslyn.Diagnostics.Analyzers/Core/SpecifyDiagnosticDescriptorCustomTags.cs]: Shall we limit this detection to fields?

File: src/Roslyn.Diagnostics.Analyzers/Core/SpecifyDiagnosticDescriptorCustomTags.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                }
+            });
+        }
+
+        private static bool IsMissingCustomTagsArgument(ImmutableArray<IArgumentOperation> arguments) =>
+            arguments.FirstOrDefault(x => x.Parameter?.Name == "customTags")?.Value is IArrayCreationOperation arrayCreation &&
```
Comment (Evangelink) [inline src/Roslyn.Diagnostics.Analyzers/Core/SpecifyDiagnosticDescriptorCustomTags.cs]: Is it possible to have a `null` parameter?

File: src/Roslyn.Diagnostics.Analyzers/Core/SpecifyDiagnosticDescriptorCustomTags.cs (hunk 1 above)
Comment (mavasani) [reply src/Roslyn.Diagnostics.Analyzers/Core/SpecifyDiagnosticDescriptorCustomTags.cs]: You don't need a new analyzer. You can just extend https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs, which already has a very similar rule for missing helpLinkUri argument.

Comment (Evangelink) [reply src/Roslyn.Diagnostics.Analyzers/Core/SpecifyDiagnosticDescriptorCustomTags.cs]: I shall pay more attention to the existing analyzers... Closing this PR and working on extending the other analyzer.

**Pull request closed without merging**